File: ios/FilmLightMeter/Domain/MeteringModels.swift
import Foundation

public enum MeteringMode: String, Codable, CaseIterable, Sendable {
    case spot
    case centerWeighted
    case average

    public var displayName: String {
        switch self {
        case .spot: "点测光"
        case .centerWeighted: "中央重点"
        case .average: "平均测光"
        }
    }
}

public enum FrameFormat: String, Codable, CaseIterable, Sendable {
    case film135
    case apsC
    case film645
    case film66

    public var displayName: String {
        switch self {
        case .film135: "135"
        case .apsC: "APS-C"
        case .film645: "6x4.5"
        case .film66: "6x6"
        }
    }

    public var sizeMillimeters: (width: Double, height: Double) {
        switch self {
        case .film135: (36, 24)
        case .apsC: (23.6, 15.7)
        case .film645: (56, 41.5)
        case .film66: (56, 56)
        }
    }
}

public struct NormalizedPoint: Codable, Equatable, Sendable {
    public let x: Double
    public let y: Double

    public init(x: Double, y: Double) {
        self.x = min(max(x, 0), 1)
        self.y = min(max(y, 0), 1)
    }
}

public struct NormalizedRect: Codable, Equatable, Sendable {
    public let left: Double
    public let top: Double
    public let right: Double
    public let bottom: Double

    public init(left: Double, top: Double, right: Double, bottom: Double) {
        precondition(left >= 0 && top >= 0 && right <= 1 && bottom <= 1)
        precondition(left < right && top < bottom)
        self.left = left
        self.top = top
        self.right = right
        self.bottom = bottom
    }

    public static let full = NormalizedRect(left: 0, top: 0, right: 1, bottom: 1)
    public var width: Double { right - left }
    public var height: Double { bottom - top }
    public var centerX: Double { (left + right) / 2 }
    public var centerY: Double { (top + bottom) / 2 }

    public func contains(x: Double, y: Double) -> Bool {
        x >= left && x <= right && y >= top && y <= bottom
    }
}

public struct MeteringConfiguration: Equatable, Sendable {
    public var mode: MeteringMode = .centerWeighted
    public var spotPoint: NormalizedPoint?
    public var spotAreaPercent = 5
    public var centerAreaPercent = 25
    public var centerWeightPercent = 70
    public var viewfinder = NormalizedRect.full
    public var previewAspectRatio = 9.0 / 16.0
    public var calibrationOffset = 0.0
    public var revision: UInt64 = 0

    public init() {}
}

public struct CameraExposureMetadata: Equatable, Sendable {
    public let exposureSeconds: Double
    public let sensitivityISO: Double
    public let aperture: Double

    public init(exposureSeconds: Double, sensitivityISO: Double, aperture: Double) {
        self.exposureSeconds = exposureSeconds
        self.sensitivityISO = sensitivityISO
        self.aperture = aperture
    }

    public var isValid: Bool {
        exposureSeconds.isFinite && exposureSeconds > 0
            && sensitivityISO.isFinite && sensitivityISO > 0
            && aperture.isFinite && aperture > 0
    }
}

public struct LuminancePlane: Sendable {
    public let width: Int
    public let height: Int
    public let values: [UInt8]
    public let isVideoRange: Bool

    public init(width: Int, height: Int, values: [UInt8], isVideoRange: Bool) {
        precondition(width > 0 && height > 0 && values.count == width * height)
        self.width = width
        self.height = height
        self.values = values
        self.isVideoRange = isVideoRange
    }
}

public struct ExposureMap: Sendable {
    public let width: Int
    public let height: Int
    public let pixelEV100: [Float]
    public let rawLuminance: [UInt8]
    public let clippedHighlights: [Bool]
    public let timestampNanoseconds: Int64
    public let revision: UInt64

    public init(
        width: Int,
        height: Int,
        pixelEV100: [Float],
        rawLuminance: [UInt8],
        clippedHighlights: [Bool],
        timestampNanoseconds: Int64,
        revision: UInt64
    ) {
        precondition(pixelEV100.count == width * height)
        precondition(rawLuminance.count == pixelEV100.count)
        precondition(clippedHighlights.count == pixelEV100.count)
        self.width = width
        self.height = height
        self.pixelEV100 = pixelEV100
        self.rawLuminance = rawLuminance
        self.clippedHighlights = clippedHighlights
        self.timestampNanoseconds = timestampNanoseconds
        self.revision = revision
    }
}

public struct ExposureSnapshot: Sendable {
    public let exposureMap: ExposureMap
    public let meteredEV100: Double
    public let measuredLuminance: Double
    public let metadata: CameraExposureMetadata
    public let timestampNanoseconds: Int64
    public let revision: UInt64

    public init(
        exposureMap: ExposureMap,
        meteredEV100: Double,
        measuredLuminance: Double,
        metadata: CameraExposureMetadata,
        timestampNanoseconds: Int64,
        revision: UInt64
    ) {
        self.exposureMap = exposureMap
        self.meteredEV100 = meteredEV100
        self.measuredLuminance = measuredLuminance
        self.metadata = metadata
        self.timestampNanoseconds = timestampNanoseconds
        self.revision = revision
    }
}

File: android/app/src/main/java/com/lightmeter/app/metering/MeteringModels.kt
package com.lightmeter.app.metering

enum class MeteringMode {
    SPOT,
    CENTER_AVERAGE,
    CENTER_CROP_AVERAGE,
    CENTER_WEIGHTED,
    AVERAGE,
}

enum class CameraMeteringPreset(
    val displayName: String,
    val meteringMode: MeteringMode,
    val meteringAreaPercent: Int?,
    val centerAreaPercent: Int,
    val centerWeightPercent: Int,
) {
    CANON_NEW_F1(
        displayName = "佳能 New F-1",
        meteringMode = MeteringMode.CENTER_WEIGHTED,
        meteringAreaPercent = null,
        centerAreaPercent = 60,
        centerWeightPercent = 70,
    ),
    CANON_AL1(
        displayName = "佳能 AL-1",
        meteringMode = MeteringMode.CENTER_WEIGHTED,
        meteringAreaPercent = null,
        centerAreaPercent = 40,
        centerWeightPercent = 65,
    ),
    OLYMPUS_35_SP_AVERAGE(
        displayName = "奥林巴斯 35 SP",
        meteringMode = MeteringMode.CENTER_AVERAGE,
        meteringAreaPercent = 20,
        centerAreaPercent = 20,
        centerWeightPercent = 95,
    ),
    OLYMPUS_35_SP_SPOT(
        displayName = "奥林巴斯 35 SP",
        meteringMode = MeteringMode.SPOT,
        meteringAreaPercent = 2,
        centerAreaPercent = 20,
        centerWeightPercent = 95,
    ),
}

data class NormalizedPoint(
    val x: Double,
    val y: Double,
) {
    init {
        require(x in 0.0..1.0) { "x must be normalized" }
        require(y in 0.0..1.0) { "y must be normalized" }
    }
}

data class NormalizedMeteringRect(
    val left: Double,
    val top: Double,
    val right: Double,
    val bottom: Double,
) {
    init {
        require(left in 0.0..1.0)
        require(top in 0.0..1.0)
        require(right in 0.0..1.0)
        require(bottom in 0.0..1.0)
        require(left < right)
        require(top < bottom)
    }

    val width: Double get() = right - left
    val height: Double get() = bottom - top
    val centerX: Double get() = (left + right) / 2.0
    val centerY: Double get() = (top + bottom) / 2.0

    fun contains(x: Double, y: Double): Boolean {
        return x in left..right && y in top..bottom
    }

    companion object {
        val Full = NormalizedMeteringRect(0.0, 0.0, 1.0, 1.0)
    }
}

data class MeteringConfig(
    val mode: MeteringMode = MeteringMode.CENTER_WEIGHTED,
    val spotPoint: NormalizedPoint? = null,
    val spotAreaPercent: Int = 5,
    val centerAverageAreaPercent: Int = 20,
    val centerCropPercent: Int = 60,
    val centerAreaPercent: Int = 25,
    val centerWeightPercent: Int = 70,
    val viewfinderRect: NormalizedMeteringRect = NormalizedMeteringRect.Full,
    val previewAspectRatio: Double = 1.0,
    val targetZoomRatio: Double = 1.0,
    val isZoomReady: Boolean = true,
    val revision: Long = 0L,
    val calibrationOffset: Double = 0.0,
) {
    init {
        require(spotAreaPercent in 1..10)
        require(centerAverageAreaPercent in 1..20)
        require(centerCropPercent in 10..100)
        require(centerAreaPercent in 5..80)
        require(centerWeightPercent in 50..95)
        require(previewAspectRatio > 0.0)
        require(targetZoomRatio > 0.0)
    }
}

data class MeteringResult(
    val ev100: Double,
    val measuredLuminance: Double,
    val timestampNs: Long,
    val revision: Long = 0L,
)

data class ExposureSnapshot(
    val exposureMap: ExposureMap,
    val meteredEv100: Double,
    val timestampNs: Long,
    val revision: Long,
) {
    init {
        require(meteredEv100.isFinite())
        require(exposureMap.timestampNs == timestampNs)
        require(exposureMap.revision == revision)
    }
}
